cmake/OpenCVCompilerOptions.cmake

```diff
@@ -377,10 +377,15 @@ if(MSVC)
 endif()
 
 if(NOT OPENCV_FP16_DISABLE)
+  if(ARM)
+    set(FP16_OPTION "-mfpu=neon-fp16")
+  elseif(X86 OR X86_64)
+    set(FP16_OPTION "-mf16c")
+  endif()
```

Comment:
@alalek , thank you for your review and suggestion.
I switched back to ```try_compile```
I was facing a problem that when the processor doesn't support FP16, the program was crashing with ``illegal instruction``
I checked and this happened even I disabled the ```CV_FP16```
I then realized that the generated binary had some AVX instructions.
```
~/opencv-fork/build$ cd modules/core
~/opencv-fork/build/modules/core$ make test/test_intrin.cpp.s
~/opencv-fork/build/modules/core$ grep vadd CMakeFiles/opencv_test_core.dir/test/test_intrin.cpp.s
        vaddps  %xmm0, %xmm1, %xmm1
        vaddps  %xmm2, %xmm3, %xmm0
        vaddps  %xmm1, %xmm0, %xmm0
```
Those ```vaddps``` instructions were not from the ```FP16``` related part.
So I let the FP16 feature rely on AVX flag, too.